src/misc/degree_sequence.cpp

```diff
@@ -68,66 +70,232 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
+            igraph_integer_t bucket = buckets_req.top();
+            buckets_req.pop();
+
+            igraph_integer_t to_get = buckets[bucket].count - num_skip;
+            while (to_get > 0) {
+                vd_pair node = buckets[bucket].nodes.top();
+                if (VECTOR(seq)[node.vertex] != 0) {                   // if "not marked for removal"
+                    IGRAPH_CHECK(igraph_vector_int_push_back(&spokes, node.vertex)); // add as spoke
+
+                    node.degree--;
+                    insert_node(node);                // first, insert into bucket below
 
-                VECTOR(*edges)[2 * (ec + i)] = vd.vertex;
-                VECTOR(*edges)[2 * (ec + i) + 1] = vertices[i].vertex;
+                    buckets[bucket].count--;
+                    to_get--;
+                }
+                buckets[bucket].nodes.pop();                  // then pop from original bucket
             }
+            num_skip = 0;
         }
+        return IGRAPH_SUCCESS;
+    }
+};
 
-        ec += vd.degree;
+enum class Method {
+    LARGEST_FIRST,
+    SMALLEST_FIRST,
+    CUSTOM
+};
+
+static igraph_error_t igraph_i_havel_hakimi(const igraph_vector_int_t *degseq,
+                                            igraph_vector_int_t *edges,
+                                            Method method,
+                                            const igraph_vector_int_t *hub_order = nullptr) {
+    igraph_integer_t n_nodes = igraph_vector_int_size(degseq);
+
+    // ----- upfront error/graphicality checks ----- //
+    if (method == Method::CUSTOM && hub_order == nullptr) {
+        IGRAPH_ERROR("Custom ordering specified but no order given", IGRAPH_EINVAL);
+    } else if (n_nodes == 0 || (n_nodes == 1 && VECTOR(*degseq)[0] == 0)) {
+        return IGRAPH_SUCCESS;
     }
 
-    return IGRAPH_SUCCESS;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        igraph_integer_t deg = VECTOR(*degseq)[i];
+        if (deg >= n_nodes) {
+            IGRAPH_ERROR("Invalid degree sequence.", IGRAPH_EINVAL);
+        }
+    }
 
-fail:
-    IGRAPH_ERROR("The given degree sequence cannot be realized as a simple graph.", IGRAPH_EINVAL);
+    // ----- main Havel-Hakimi loop ----- //
+    // O(V + alpha(V) * E)
+    // O(V + E) for the LARGEST_FIRST method
+    igraph_vector_int_t seq;
+    IGRAPH_CHECK(igraph_vector_int_init_copy(&seq, degseq));
+
+    HavelHakimiList vault(seq);
+
+    igraph_integer_t n_edges_added = 0;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        // hub node selection
+        vd_pair hub;
+        if (method == Method::SMALLEST_FIRST) {
+            if (!vault.get_min_node(/* out param */hub)) break;
+            vault.remove_min_node();
+        }
+        else if (method == Method::LARGEST_FIRST) {
+            if (!vault.get_max_node(/* out param */hub)) break;
+            vault.remove_max_node();
+        }
+        else if (method == Method::CUSTOM) {
```

Comment:
`IGRAPH_REALIZE_DEGSEQ_INDEX` still uses the old implementation. The justification for using the more complex linked-list based method was to be able to support this case with good complexity, as we already have simpler approachers for the LARGEST and SMALLEST cases.

Let's have a performant implementation for `IGRAPH_REALIZE_DEGSEQ_INDEX` as well before merging this PR.